### ISSUE TITLE
Using a launcher to launch a script which invokes further launch commands

### ISSUE THREAD

**Issue by dsolt** [pmix/pmix#1211]
Thank you for taking the time to submit an issue!

## Background information

master from git clone on power systems, TCP

-----------------------------

## Details of the problem

When processes are created by a launcher, that launcher is directed to call PMIx_server_setup_fork() to setup an environment and PMIx_server_register_client() to inform the convenience library of the process being created.  We do this so that the process being created can call PMIx_Init and have in its environment information on how to connect to the pmix server.   However, it is often the case that the thing being launched is not a PMIx client, but instead is a script containing commands to launch further applications.   For example:

`
% cat myjob

#!/bin/sh
mylauncher --np 16 a.out
mylauncher --np 64 b.out

% mylauncher --np 1 myjob`

I expect that many implementations of a launching system would implement mylauncher using a call to PMIx_tool_init().   It seems to me that these tools should not be influenced by their environment, but that they should always connect in the same way.  They are tools and not clients and shouldn't be looking in their environment, which is probably not intended for them.  An easy solution to this is to have mylauncher unset all relevant PMIX environment variables that are not meant for it.  If this is the best way, note that it is not a maintenance free approach since someone might introduce another connection method with new environment variables tomorrow, making the current list of unset's insufficient.  The other approach seems to be that PMIx_tool_init does not look in the environment for how to connect.  You have to pass in info's if you want to connect to the pmix server in a specific way.   It seems like this is the better approach.  What do others think?

**Comment by rhc54** [comment]
I'm not sure I understand, so let me probe a bit here. We've been doing a bunch of work on exactly this scenario for the tool vendors, so maybe first explaining a bit about how this is intended to work might help.

Note: the tool support chapter is being written, but isn't yet complete. Hopefully, we'll get it done in May so this scenario is easier to grok.

When a tool (e.g., your `mylauncher`) intends to act as a launcher, it indicates that in the attributes passed to PMIx_tool_init. This has some very specific consequences as to how the tool will behave (e.g., it initializes the PMIx server portions of the library so other tools can connect to it). However, it still allows for the tool to connect back to the server that launched it (assuming it was started that way vs just at a cmd line) unless another attribute directs it not to do so.

This is actually a desirable outcome. `mylauncher` may, for example, want to use the system PMIx server to launch its applications. Because it is in fact a client of the system PMIx server, it has to connect to that server as an official `client`. Note that the system PMIx server doesn't necessarily know that `mylauncher` is going to initialize as a tool and/or launcher, so it can't know not to include connection info in the environment.

My guess is that you're running into trouble because `mylauncher` is copying its environment and passing it to its child processes, and it is doing so without itself calling `PMIx_server_setup_fork` (otherwise, those envars would be overwritten by the new ones). Is that an accurate description?

You could, of course, direct `mylauncher` to "not connect" to its server, but that won't help the environment problem. Neither will having the tool not look in the environment for connection info - the connection info will still be there and cause trouble with the ensuing clients.

If environment "pollution" for the subsequent launched procs is the issue here, then the easiest solution might be for `mylauncher` to simply remove all envars starting with `PMIX_` that do not start with `PMIX_MCA`. Alternatively, you could ensure that `mylauncher` calls `PMIx_server_setup_fork` so the envars are correctly overwritten.

Can you clarify that environment "pollution" is the problem you are trying to solve?

**Comment by dsolt** [comment]
My guess is that the right answer here is that in mylauncher, when it calls PMIx_tool_init, it passes in info arguments that direct how to connect to the server instead of taking the default (which in this case is to use the environment which was setup for another client, not the tool itself).   

In my case, the PMIx server is launching a process which it thinks is a client, but in fact it not.  It is actually just a script  that runs several instances of mylauncher (a PMIX tool, not a traditional client).   However, mylauncher is trying to connect using the environment information that was intended for the client (which sort of works ok), but then when a 2nd mylauncher invocation occurs, the server denies it since it is not expecting another client to connect using the connection information it gave out.   

So, to summarize the environment pollution is NOT that mylauncher is polluting the environment of its children, it is that it's own environment is polluted because it is running in an environment that was intended for a single client  but is actually executing a script issuing multiple tools, each of which is trying to connect to the server using the environment intended for a single client.   

My initial reaction is that PMIx_tool_init should always direct specifically how to connect to the server and should never be influenced by its environment since it really has no idea what environment it may be running inside of.

**Comment by rhc54** [comment]
As I said, the problem is that there are use-cases where the tool has to connect as a client in order to ensure that it is connecting to the correct server. Remember, you might well be on a node where there are multiple servers running (e.g., have your "mylauncher" invocations execute in parallel) - there has to be some method for telling the tool how to select the correct target, and the environment is the only thing available to us.

I'm a little surprised by the scenario you describe as the ref implementation (PRI), at least, has no issue with multiple processes connecting with the same connection info and (nspace,rank) identifier. However, your example is a little different in that the "clones" are appearing sequentially and not in parallel - I suspect the issue, therefore, is that the PRI isn't properly accounting for a client that "finalizes" and then "reappears".

That's a use-case we simply missed - effectively a "bug" that needs to be fixed as it certainly will be seen. I should have some time next week to poke into it a bit, if someone else hasn't solved it by then.

**Comment by dsolt** [comment]
Hmmm. I think I see what you are saying.   I didn't think about the case where there are multiple servers running and that is a good point.  (Our current connection method only works for a single server per user per node and that is going to lead to problems eventually). So I agree it would be better that they do connect back on the "expected" connection as they find it in their environment.  Perhaps this problem is just a bug in the finalize/reappear case as you mentioned.   We do PMIx_Finalize() each tool normally.

**Comment by rhc54** [comment]
Sorry for delay - I was on travel all week and just got back. I think that is correct, but you've made me go back and work thru it all again...and there may be an issue here that bears further consideration. Specifically, consider your example scenario. If each `mylauncher` spawns a job using the RM, then the PMIx server in the RM is going to see the same (nspace,rank) issuing the spawn request each time. According to the standard, the RM is required to generate a "job terminated" event when the job completes, and that event is to be "targeted" to the requestor's ID.

This works okay when the spawns occur sequentially, but now consider the case where we execute the `mylaunchers` in parallel. Each `mylauncher` will have the identical ID - and so they are all going to receive the termination event for each job. The event includes the nspace of the terminating job, so the `mylauncher` tool can, if code correctly, check the event and ignore those from jobs other than the one it started. However, if they instead just thought that the RM would only notify them about their own job...well, then they will be wrong.

I think that is still okay, but we probably need to make that clear somewhere? This "multi-launcher in a script" scenario is tricky and I can see how/why people could easily get it wrong.
